### PR TITLE
Ports: Add OpenJDK Port with ZeroVM

### PR DESCRIPTION
**Pull request by ADKaster**
Port of OpenJDK 17.0.2, zero VM only.
Original porting work done by @TSultanov , reorganized and cleaned up by me.

More work needed to get the full hotspot VM up and running :^)

I verified this works on x86_64 GNU and x86_64 Clang.

Screenshot from 2022-06-14 19-49-05

### CODE REVIEW

File: Ports/OpenJDK/package.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+version="17.0.2"
+workdir=jdk17u-dev-jdk-17.0.2-ga
+useconfigure=true
+use_fresh_config_guess=true
+config_guess_paths=("make/autoconf/build-aux/autoconf-config.guess")
+use_fresh_config_sub=true
+config_sub_paths=("make/autoconf/build-aux/autoconf-config.sub")
+auth_type=sha256
+files="https://github.com/openjdk/jdk17u-dev/archive/refs/tags/jdk-17.0.2-ga.tar.gz jdk-17.0.2-ga.tar.gz cb5b2a5d0916723d340f2c5bacd4607f8b8dc3a18dc8019fcfabf5306e2a4112"
+depends=("fontconfig" "libffi")
```
Comment (linusg) [inline Ports/OpenJDK/package.sh:13]: We're going to aggressively enforce sane quoting of bash variables in ports scripts going forward :)

- Everything must be quoted, as everything is treated as strings
- If it doesn't do variable substitution, use single quotes

File: Ports/OpenJDK/package.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+port="OpenJDK"
+version="17.0.2"
+workdir=jdk17u-dev-jdk-17.0.2-ga
+useconfigure=true
+use_fresh_config_guess=true
+config_guess_paths=("make/autoconf/build-aux/autoconf-config.guess")
+use_fresh_config_sub=true
+config_sub_paths=("make/autoconf/build-aux/autoconf-config.sub")
+auth_type=sha256
+files="https://github.com/openjdk/jdk17u-dev/archive/refs/tags/jdk-17.0.2-ga.tar.gz jdk-17.0.2-ga.tar.gz cb5b2a5d0916723d340f2c5bacd4607f8b8dc3a18dc8019fcfabf5306e2a4112"
```
Comment (linusg) [inline Ports/OpenJDK/package.sh:12]: Please try to use `${version}` as much as possible

**Review by linusg — COMMENTED** [review]
sweet. I'm not going to review 4k lines of patches though, I trust you they're good :)

File: Ports/OpenJDK/package.sh (hunk 1 above)
Comment (ADKaster) [reply Ports/OpenJDK/package.sh]: Fixed quotes as discussed on discord

File: Ports/OpenJDK/package.sh (hunk 2 above)
Comment (ADKaster) [reply Ports/OpenJDK/package.sh]: good point. for major versions we'll still have to edit the path here.

**Review by timschumi — COMMENTED** [review]
Looking great, although I gave the patches only a cursory look.

A few small points of clarification/nitpicking remain.

File: Ports/OpenJDK/package.sh [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        --with-extra-cxxflags="${WARNING_IGNORE_FLAGS}" \
+        "${TOOLCHAIN_ARGS[@]}"
+}
+
+build() {
+    run make java.base jdk.compiler java.logging
+}
+
+install() {
+    run sh -c "cp ./build/serenity-* ${SERENITY_INSTALL_ROOT}/usr/Ports/ -rf"
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Ports/OpenJDK/package.sh:57]: hm yeah moving all these guys to /usr/local/lib/jvm seems reasonable. it'll still have the serenity-blah/jdk path though, at least until we get "make image" or whatever the proper install step is working to get a separate jre and jdk

File: Ports/.hosted_defs.sh [hunk 4]
```diff
@@ -23,24 +24,28 @@ export SERENITY_SOURCE_DIR="$(realpath "${SCRIPT}/../")"
...
     export CC="clang --target=${SERENITY_ARCH}-pc-serenity --sysroot=${SERENITY_BUILD_DIR}/Root"
     export CXX="clang++ --target=${SERENITY_ARCH}-pc-serenity --sysroot=${SERENITY_BUILD_DIR}/Root"
     export AR="llvm-ar"
     export RANLIB="llvm-ranlib"
     export READELF="llvm-readelf"
     export OBJCOPY="llvm-objcopy"
     export STRIP="llvm-strip"
-    export PATH="${SERENITY_SOURCE_DIR}/Toolchain/Local/clang/bin:${HOST_PATH}"
+    export CXXFILT="llvm-cxxfilt"
+    export PATH="${SERENITY_TOOLCHAIN_BINDIR}:${HOST_PATH}"
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Ports/.hosted_defs.sh]: Fair nuff, moved down

File: Userland/Libraries/LibC/grp.cpp [hunk 5]
```diff
@@ -59,19 +59,25 @@ struct group* getgrgid(gid_t gid)
...
     setgrent();
     while (auto* gr = getgrent()) {
-        if (gr->gr_gid == gid)
+        if (gr->gr_gid == gid) {
+            endgrent();
             return gr;
+        }
     }
+    endgrent();
     return nullptr;
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Userland/Libraries/LibC/grp.cpp]: True. See C, write C. Changed to use a scopeguard.

File: Ports/OpenJDK/patches/0007-java.base-Update-native-modules-to-support-Serenity.patch [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
++#endif
++
+ /* To be able to name the Java constants the same as the C constants without
+    having the preprocessor rewrite those identifiers, add PREFIX_ to all
+    identifiers matching a C constant. The PREFIX_ is filtered out in the
+@@ -50,6 +139,7 @@ class UnixConstants {
+     static final int PREFIX_O_TRUNC = O_TRUNC;
+     static final int PREFIX_O_SYNC = O_SYNC;
+ 
++
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Ports/OpenJDK/patches/0007-java.base-Update-native-modules-to-support-Serenity.patch]: deleted, nice eye, I completely ignored these when reviewing.

File: Ports/OpenJDK/package.sh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    WARNING_IGNORE_FLAGS='-Wno-error=switch -Wno-maybe-uninitialized'
+    if [ $SERENITY_TOOLCHAIN = 'Clang' ]; then
+        # We need the build CC and CXX to actually be clang when using clang to cross-compile
+        #    ... for some reason.
+        TOOLCHAIN_ARGS=("--with-toolchain-type=clang"
+                        "BUILD_CC=clang"
+                        "BUILD_CXX=clang++")
+        WARNING_IGNORE_FLAGS="${WARNING_IGNORE_FLAGS} -Wno-error=bitwise-instead-of-logical"
+    else
+        :
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Ports/OpenJDK/package.sh]: Fair enough, removed the placeholder. we'll add when needed.

**Pull request merged**